kvclient.go

```diff
@@ -202,33 +209,60 @@ func NewKvClient(ctx context.Context, opts *KvClientConfig) (*kvClient, error) {
 	return kvCli, nil
 }
 
-func (c *kvClient) Reconfigure(ctx context.Context, opts *KvClientConfig) error {
-	if opts == nil {
-		return nil
+func (c *kvClient) Reconfigure(config *KvClientConfig, cb func(error)) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if config == nil {
+		return errors.New("must specify a configuration to reconfigure to")
 	}
 
-	if opts.TlsConfig != c.tlsConfig {
-		return placeholderError{"cannot reconfigure tls config"}
+	if c.currentConfig.Address != config.Address ||
+		c.currentConfig.TlsConfig != config.TlsConfig ||
+		c.currentConfig.ClientName != config.ClientName ||
+		c.currentConfig.Authenticator != config.Authenticator ||
+		c.currentConfig.DisableDefaultFeatures != config.DisableDefaultFeatures ||
+		c.currentConfig.DisableErrorMap != config.DisableErrorMap ||
+		c.currentConfig.DisableBootstrap != config.DisableBootstrap {
+		// pretty much everything triggers a reconfigure
+		return errors.New("cannot reconfigure due to conflicting options")
 	}
-	if opts.Address != c.hostname {
-		return placeholderError{"cannot reconfigure address"}
+
+	var selectBucketName string
+	if config.SelectedBucket != c.currentConfig.SelectedBucket {
+		if c.currentConfig.SelectedBucket != "" {
+			return errors.New("cannot reconfigure from one selected bucket to another")
+		}
+
+		// because we only support going from no selected bucket to a selected
+		// bucket, we simply update the state here and nobody will be permitted to
+		// reconfigure unless it fails and set its back to no selected bucket.
+		c.currentConfig.SelectedBucket = config.SelectedBucket
+		selectBucketName = config.SelectedBucket
 	}
-	if opts.Authenticator != c.authenticator {
-		return placeholderError{"cannot reconfigure authenticator"}
+
+	if !c.currentConfig.Equals(config) {
+		return errors.New("client config after reconfigure did not match new configuration")
 	}
 
-	if opts.SelectedBucket != "" {
-		if c.bucket != "" {
-			return placeholderError{"cannot perform select bucket on an already bucket bound kvclient"}
+	go func() {
+		if selectBucketName != "" {
+			err := c.SelectBucket(context.Background(), &memdx.SelectBucketRequest{
+				BucketName: config.SelectedBucket,
```

Comment:
Should probably be selectBucketName, that or make selectBucketName a bool instead.

Reply:
Fixed.